lib/src/Datepicker/Datepicker.tsx

```diff
@@ -139,18 +142,36 @@ export const Datepicker: React.FunctionComponent<DatepickerProps> = React.forwar
         };
 
         React.useEffect(() => {
-            const day: number = monthPicker ? 1 : customDay;
-            const month: number = customMonth;
-            const year: number = customYear;
-            const dateString: string = `${padNumber(year, true)}-${padNumber(month)}-${padNumber(day)}`;
-            const date: Date = new Date(dateString);
-            const m: number = date.getMonth() + 1;
-            if (date.getFullYear() === year && m === month && date.getDate() === day) {
-                isDateInRange(date, min, max) ? onChange(date) : onChange(null);
-            } else {
-                onChange(null);
+            setRenderType(() => {
+                if (forceCustom) {
+                    return "custom";
+                }
+                if (monthPicker && supportsInputOfType("month")) {
+                    return "month";
+                }
+                if (supportsInputOfType("date")) {
+                    return "date";
+                }
+                return "custom";
+            });
+        }, [forceCustom, monthPicker]);
+
+        React.useEffect(() => {
+            if (renderType === "custom") {
+                // only call onchange when the field is custom datepicker
+                const day: number = monthPicker ? 1 : customDay;
+                const month: number = customMonth;
+                const year: number = customYear;
+                const dateString: string = `${padNumber(year, true)}-${padNumber(month)}-${padNumber(day)}`;
+                const date: Date = new Date(dateString);
+                const m: number = date.getMonth() + 1;
+                if (date.getFullYear() === year && m === month && date.getDate() === day) {
+                    isDateInRange(date, min, max) ? onChange(date) : onChange(null);
+                } else {
+                    onChange(null);
```

Comment:
from the code, it seems like `onChange` is still being fired on render here

Reply:
i was thinking to skip this, but added another commit for this